### ISSUE TITLE
Vector Tiles

### ISSUE THREAD

**Issue by pjcozzi** [CesiumGS/cesium#2132]
2D maps like Google, Apple, MapBox GL, and OL3 are moving to actual vectors for rendering vector data, instead of baking vector data into a raster tile.  This allows, for example, to make labels always face up even if south is up and to fade out labels so they never hit each other.

For now, this roadmap is a collection of resources on this topic.
- [Google Maps + WebGL = MapsGL](https://docs.google.com/presentation/d/1pL-OZnsWIKWcXflfJV8MPH3kKBehn0Mlx6-cdjmP3A0/edit#slide=id.g1e3a90e_1_28)
- [Vector tiles for fast custom maps](http://vimeo.com/106228141) (video)
- [Videos from FOSS4G](https://www.mapbox.com/blog/foss4g-jsgeo-videos/)
- [Mapbox Vector Tile Specification](https://github.com/mapbox/vector-tile-spec)
  - [blog post](https://www.mapbox.com/blog/vector-tiles/)
  - tools: [vector-tile-js](https://github.com/mapbox/vector-tile-js) and [mapnik-vector-tile](https://github.com/mapbox/mapnik-vector-tile)
  - Client-side: pbf.js to decode tiles (protocol buffers).
- [geojson-vt](https://github.com/mapbox/geojson-vt)
- [MMT-Vector-Tiles](https://github.com/glob3mobile/mmt-vector-tiles/)

**Comment by pjcozzi** [comment]
More discussion: https://groups.google.com/forum/#!topic/cesium-dev/QnOxsA0iwa0

**Comment by mramato** [comment]
Looks like ESRI is adopting Mapbox vector tiles: https://www.mapbox.com/blog/vector-tile-adoption/

**Comment by pjcozzi** [comment]
Info from [FOSS4G NA 2015](http://geothought.blogspot.com/2015/03/review-of-foss4g-na-2015.html).

**Comment by pjcozzi** [comment]
From Mapzen: [Look Upon Our Squares of Math in Three Dimensions](https://mapzen.com/blog/look-upon-our-squares-of-math-in-three-dimensions)

**Comment by pjcozzi** [comment]
* State of the Map 2015 - Vector Rendering Panel - [video](https://www.youtube.com/watch?v=0RA5k7J5Sw0), [transcript](https://s3.amazonaws.com/assets-staging.mapzen.com/images/state-of-the-map-2015-wrapup/sotmus2015_Mapzen_transcripts.html#matt)

**Comment by pjcozzi** [comment]
http://osgeo-org.1560.x6.nabble.com/Asking-for-new-vector-tiles-community-module-tt5211994.html

**Comment by pjcozzi** [comment]
* From Esri: http://blogs.esri.com/esri/arcgis/2015/07/20/vector-tiles-preview/

**Comment by GatorScott** [comment]
Demo of vector tiles at Esri UC is last video on page

http://www10.giscafe.com/blogs/gissanjay/2015/07/20/keynote-from-jack-dangermond-at-the-2015-esri-user-conference/

**Comment by bensleveritt** [comment]
Is this still being considered by Cesium? It would be an impressive feature to tout.

**Comment by pjcozzi** [comment]
Yes, but our current focus is on [3D Tiles](https://github.com/AnalyticalGraphicsInc/3d-tiles).

**Comment by bingqixuan** [comment]
Hello, I want to know when will it come out, I'm very looking forward to it~

**Comment by pjcozzi** [comment]
Current work-in-progress for 3D Tiles vector tiles is in the [vector-tiles](https://github.com/AnalyticalGraphicsInc/cesium/tree/vector-tiles) branch.  The final date is still TBA.

**Comment by pjcozzi** [comment]
@davidyaha that added support for the draft vector tile payload in 3D Tiles, but I would still leave this issue open for other potential vector tile specs.

**Comment by OmarShehata** [comment]
OSM has global vector data, it would be cool to support that in Cesium and eventually serve it from ion:

https://openmaptiles.com/downloads/dataset/osm/#0.23/0/-26

**Comment by OmarShehata** [comment]
Requested again here: https://github.com/AnalyticalGraphicsInc/cesium/issues/7939

Also perhaps relevant, some thoughts on how TerriaJS loads these vector tiles? https://github.com/AnalyticalGraphicsInc/cesium/issues/6182#issuecomment-362723885

**Comment by bampakoa** [comment]
Is there any related work or PR that we could follow up?

**Comment by icemagno** [comment]
How can I extend `MapboxImageryProvider` to load PBF files instead PNG ? I think I can do something to put is to work using this : https://landtechnologies.github.io/Mapbox-vector-tiles-basic-js-renderer/.

As I can see, this renderer can get a PBF from `{x},{y},{z}` and render it in any canvas (the Cesium canvas??).

it is just an idea to start. I know almost nothing about how Providers work.

What do you say?

**Comment by OmarShehata** [comment]
@icemagno you might find the GridImageryProvider useful as a reference here. It's the simplest imagery provider I know of (procedural creates a grid as the tiles): https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/GridImageryProvider.js

It should be straightforward to set up the CesiumJS source code so you can add your custom class, see the build guide here: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/BuildGuide/README.md#build-guide

**Comment by icemagno** [comment]
Ok. I'll give a try.
BTW: I know you guys have a long expertise to write thousands of lines before build and test but I like to check every step I do in the code.

So... I dont like the "write-build-test" method because I'll spend hours in a lot of builds. Is there some "write-F5-test" way to do this? I mean... can I write my code and immediately check it just refreshing the browser without the need to rebuild every time?

**Comment by OmarShehata** [comment]
> BTW: I know you guys have a long expertise to write thousands of lines before build and test 

I don't think that's a good idea no matter how much expertise anyone has :smile:

When running CesiumJS from the source as described in the linked guide above, you can make changes directly to the source and see the changes immediately by refreshing the page or re-running the Sandcastle example. This is how most CesiumJS development is done. 

If you have any follow up questions on best practices here, let's take this conversation [to the forum](https://groups.google.com/forum/#!forum/cesium-dev) so as not to derail this GitHub issue.

**Comment by am2222** [comment]
> How can I extend `MapboxImageryProvider` to load PBF files instead PNG ? I think I can do something to put it to work using this : https://landtechnologies.github.io/Mapbox-vector-tiles-basic-js-renderer/.
> 
> As I can see, this renderer can get a PBF from `{x},{y},{z}` and render it in any canvas (the Cesium canvas??).
> 
> https://landtechnologies.github.io/Mapbox-vector-tiles-basic-js-renderer/debug/basic/google
> 
> it is just an idea to start. I know almost nothing about how Providers work. I need to know how to test a new Provider without need to compile all Cesium code. Any way to add it as a plugin? Some code to show me how?
> 
> What do you say?

@icemagno What if we suse something like this library
https://github.com/Leaflet/Leaflet.VectorGrid

**Comment by icemagno** [comment]
@am2222 I'm using Cesium not Leaflet

**Comment by am2222** [comment]
@icemagno yeah I know, I am also looking for a way to make vector tiles work on cesium, but that plugin actually is doing the same thing as you said.it reads pbf files draws them on a canvas and leaflet plots that canvas. in our case we need draw pbf using cesium's canvas.

**Comment by nf-s** [comment]
Terriajs has a MapboxVectorTileImageryProvider which draws (polygons only at this point in time) to canvas - see https://github.com/TerriaJS/terriajs/blob/master/lib/Map/MapboxVectorTileImageryProvider.js
@steve9164 created this a few years ago, he can give you more info about it

**Comment by am2222** [comment]
@nf-s I think the provided sample helps me alot. I have not looked at it in details yet, Does it use any internal files fron terriajs library as well?

**Comment by nf-s** [comment]
It uses 2 small utility functions from the terria library which are essential, but everything else is imported from Cesium, @mapbox/vector-tile, pbf and urijs (except the i18next internationalisation stuff which can be stripped out)

**Comment by am2222** [comment]
@nf-s thanks I could make it work

**Comment by davenquinn** [comment]
Either of you make any progress on this with working code? I was going to try to hack together an implementation soon as well.

**Comment by am2222** [comment]
@davenquinn this code works well with polygons
https://github.com/TerriaJS/terriajs/blob/master/lib/Map/MapboxVectorTileImageryProvider.js

**Comment by robbo1975** [comment]
Hi,  I am currently working an an implementation of a StyledVector Map tile provider for Cesium.  I am doing this work for a commercial project that I am working on and am willing to contribute the code to the community project. I am making steady progress and will be continuing to develop a solution until it provides at least around 85% of the StyledVector map features.  Would the community be interested in receiving this code and if so should I submit through a pull request on a branch associated with this ticket?
![image1](https://user-images.githubusercontent.com/64151514/83286359-a9617b00-a1d7-11ea-85dc-08f4a8421ea1.png)
![image2](https://user-images.githubusercontent.com/64151514/83286378-b54d3d00-a1d7-11ea-92c2-e291e0ccbd02.png)

**Comment by am2222** [comment]
@robbo1975 I think it would be great thing having vectoretiles in cesium,

**Comment by davenquinn** [comment]
@robbo1975: excited to see this contribution! It is sorely needed.

There are different strategies taken by different providers for styling vector layers (e.g. Mapbox, Carto, Deck.gl, and probably OpenLayers although I haven't looked deeply). Does your layer take a separate strategy? Or does it adopt one of the existing solutions (e.g. CartoCSS)?

Either way, would it make sense to first package the project outside of core Cesium as a separate module? This might allow it to evolve quickly and independently — if it proves to be a widely useful solution, it could easily be integrated into Cesium itself at that point. I recently packed up a [terrain provider](https://github.com/davenquinn/cesium-martini) that might be a good example of how to structure an external Cesium plugin.

**Comment by robbo1975** [comment]
@davenquinn I have adopted the mapbox standard in my approach because of a need to support ESRI Vector tile services. Your suggestion of developing as an external plugin module seems eminently sensible; however, JavaScript is not my first language so I'm still trying to find my feet with all of the language concepts and associated Frameworks and tools! I will take a look at your example module and see if I can replicate the core modular capability. Thanks for your feedback :)

**Comment by davenquinn** [comment]
@robbo1975 the Mapbox standard is the most widely used, so having a renderer targeted at that would be wonderful, and extremely useful for our work as well. If you were able to create a project, I'd be happy to help with code review etc. (as someone who is mostly an outsider to Cesium, but a fairly proficient javascript developer).

**Comment by kring** [comment]
Hey @robbo1975, thanks for working on this and for sharing your code! I just wanted to mention real quick that CesiumJS can't accept GPL-licensed code because of CesiumJS's much more permissive Apache 2.0 license. I suspect most CesiumJS applications would also be unable to use it. Not sure if the license was a conscious choice there, but thought I should mention it.

**Comment by robbo1975** [comment]
@kring Thanks for the observation; it was an oversight on my part.  I have updated the license to the more permissive MIT license

**Comment by hmartiro** [comment]
+1 for Mapbox vector tile support!

**Comment by TJPovey** [comment]
has there been any further development on this? There are several vector basemaps provided by esri that I would love to use as imagery providers

**Comment by robbo1975** [comment]
Hi,

Thanks for the interest. I do need to complete this project, but recently
I've been tied up with other work. I will need to complete within the next
6 months or so, but with will depend on getting time to focus on the
improvements that need to be made.

Happy to take assistance if anybody has the skills to help.

Robbo

**Comment by kikitte** [comment]
Hi, here is another implementation of MapboxVectorTileImageryProvider using Mapbox-vector-tiles-basic-js-renderer, https://github.com/kikitte/MVTImageryProvider.  I hope it will help somebody who want to render mapbox vector tile in cesium.

**Comment by davenquinn** [comment]
Hi — just an update on this thread: I tried several approaches to this rendering problem for my work on [visualizing Mars](https://dev.macrostrat.org/mars/syrtis-jezero/layers#a=328&e=50&overlays=%2Cgeology&x=78.345&y=15.473&z=140093) in advance of the Perseverance landing this week. The approach brought over from TerriaJS works well but is relatively slow. @kikitte's renderer based on the Mapbox GL fork also works well and is much quicker (that's what I ended up using, ultimately). I think some sort of WebGL approach based on Mapbox GL (pre-license-change) or Deck.gl would be appropriate long term, with the significant benefit that we could reuse some of the work on styling/symbology that has gone into these renderers. There's no reason why we couldn't have multiple vector-tile backends, too.

<img width="1431" alt="Screen Shot 2021-02-18 at 12 17 40 PM" src="https://user-images.githubusercontent.com/1537910/108607566-be45f600-7386-11eb-8abb-8df80516f3c8.png">

**Comment by ChillAndImprove** [comment]
Updates?

**Comment by davenquinn** [comment]
Actually, yeah there are some updates! I just finished the initial public rollout of the [cesium-vector-provider module](https://github.com/davenquinn/cesium-vector-provider) and created a [demo website](https://davenquinn.com/viz/cesium-vector-provider/) that shows it in action.

I think this will be pretty great implementation that maximizes compatibility with the Mapbox stack, especially once some of the outstanding design deficiencies are addressed. I'm working to ensure that the project can be based directly on the rendering stack within Maplibre GL (see https://github.com/maplibre/maplibre-gl-js/issues/166). It'd be nice if we could eventually drop in a Mapbox stylesheet to Cesium and get a functional basemap.

Contributions to the project are welcome!

**Comment by davenquinn** [comment]
Hi @ebogo1, and thanks for looking at this. I agree that the size of Maplibre GL is a bit of a problem here. Right now the `cesium-vector-provider` module references a full/compiled version of Maplibre, but that is because I haven't figured out the right strategy for referring to things directly from their source code (a large part of the problem here is replicating their wasm compilation pipeline). This is probably simple but the module just hasn't evolved to that point.

Thinking ahead, I suspect that having this component as a standalone module makes sense. Ideally, it would just be a custom layer that can be imported with all dependencies from an external node module (this is the role that `cesium-vector-provider` seeks to fill). This would allow glue code between Maplibre and Cesium to have a lifecycle independent of Cesium's internals, and would allow (the large majority of) people with no need for Mapbox-compatible vector rendering to omit this heavy dependency.

I don't know if keeping things outside of the Cesium core library is a preferred approach of the team or how it fits with your development workflow, but it's how I'd try to proceed, at least initially.

**Comment by davenquinn** [comment]
I will also note that one of the potential improvements to rendering will be if we can directly render into Cesium-managed canvas contexts rather than creating our own, serializing as images, and exposing in a tile provider.

This feels like a lot of hoops to jump through and is purely a function of my and previous folks’ unfamiliarity with Cesium’s internals.

**Comment by nf-s** [comment]
A light-weight alternative to mapbox-gl/maplibre-gl is https://github.com/protomaps/protomaps.js - it supports MVT and a subset of mapbox style spec.

We have been using it in TerriaJS for nearly a year - and it works quite well.

https://github.com/TerriaJS/terriajs/blob/main/lib/Map/ImageryProvider/ProtomapsImageryProvider.ts

**Comment by gberaudo** [comment]
For information, MapLibre is building their roadmap and calling for popular vote.
There is a proposition for a minimal build which may align with your needs @davenquinn.
See https://github.com/maplibre/maplibre/discussions/159 / https://maplibre.org/news/2023-03-13-second-call-for-bounties/

**Comment by puzonevan** [comment]
@davenquinn Any updated progress on the module?

**Comment by markchagers** [comment]
We really look forward to vector tile (PBF) support for cesium. The newer ESRI services use this format, and currently cannot be used in cesiumjs. This is a much needed enhancement.

**Comment by DoisKoh** [comment]
Fyi we're in a similar boat and our company is having a hard time justifying going all-in with Cesium because of this.

**Comment by markchagers** [comment]
Is there any news regarding this feature? The Esri dev team is prodding us to switch to the newer vector tile service, but we can't because of lacking support in CesiumJS. Please don't force us to go the hard way of replacing Cesium with mapLibre in our project!

**Comment by kteneyck** [comment]
I'm looking to render large 2D datasets like all US airspace, railroads lines, power transmission lines, etc. Right now it needs to be rendered as multiple Layers which results in a lot of png's streamed to the application

**Comment by davenquinn** [comment]
Reminder here that community contributions to https://github.com/davenquinn/cesium-vector-provider would move this capability forward meaningfully. Updating to a newer version of MapLibre GL JS, particularly, would allow more flexibility, but I am sure there are also tons of efficiency gains that can be realized.

I plan to put more effort into this module at some point but my priorities are elsewhere for now.

**Comment by MasahiroWatanabeEB** [comment]
Hi, I would like to use ESRI vector map service on cesium.js  I'd appreciate if anyone could contribute to move forward...

**Comment by ggetz** [comment]
Since this issue contains many requests for MVT support, I'm cross-linking @danielzhong's [PR introducing MVT loading](https://github.com/CesiumGS/cesium/pull/13404) 🎉

**Comment by jo-chemla** [comment]
Just had a quick look at the linked PR, thanks for sharing, seems that there is great progress on the integration, congrats to the author and cesium team!

According to this comment https://github.com/CesiumGS/cesium/pull/13404#discussion_r3202406301 it looks like the current implementation leaves room for a somewhat easy path to extend this MVT implementation to adding also support for MLT, Maplibre Tiles see [spec](https://maplibre.org/maplibre-tile-spec/).  They ship with an existing implementation of a [ts decoder](https://github.com/maplibre/maplibre-tile-spec/tree/main/ts) which will probably help decode tiles payload. 

This thread tries to gather current ongoing integrations and existing datasets https://github.com/maplibre/maplibre-tile-spec/issues/1387. MLT is currently expanding outside the pure maplibre ecosystem, with planetiler, pmtiles, soon tippecanoe output implementations. Example [osm-dataset](https://onthegomap.github.io/planetiler-demo/#9.75/41.5031/-72.3876) produced by onthego + planetiler, see below the note for the added benefits of MLT over MVT

> - Improved compression ratio - up to 6x on large tiles, based on a column oriented layout with (custom) lightweight encodings
> - Better decoding performance - fast lightweight encodings which can be used in combination with SIMD/vectorization instructions
> - Support for linear referencing and m-values to efficiently support the upcoming next generation source formats such as Overture Maps (GeoParquet)
> - Support 3D coordinates, i.e. elevation
> - Support complex types, including nested properties, lists and maps
> - Improved processing performance: Based on an in-memory format that can be processed efficiently on the CPU and GPU and loaded directly into GPU buffers partially (like polygons in WebGL) or completely (in case of WebGPU compute shader usage) without additional processing
